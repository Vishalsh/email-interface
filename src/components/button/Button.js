import React from 'react';
import PropTypes from 'prop-types';

import classes from './Button.module.scss';

const Button = (props) => {
  const {
    children,
    type,
    onClick,
    disabled,
    className,
    ...others
  } = props;

  return (
    <button className={`${classes.button} ${classes[type]} ${className}`} onClick={ onClick } disabled={disabled} { ...others }>
      { children }
    </button>
  );
};

Button.propTypes = {
  type: PropTypes.string,
  disabled: PropTypes.bool,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  type: 'default',
  disabled: false,
  className: ''
};

export default Button;
