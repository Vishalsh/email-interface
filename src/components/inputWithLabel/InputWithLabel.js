import React from 'react';
import PropTypes from 'prop-types';

import classes from './InputWithLabel.module.scss';

const InputWithLabel = (props) => {
  const {
    label,
    value,
    inputType,
    textarea,
    className,
    placeholder,
    onChange
  } = props;

  return (
    <div className={`${classes.inputWithLabel} ${className}`}>
      <label className={classes.label}
             htmlFor={label}>
        {label.toUpperCase()}
      </label>
      {
        textarea ?
          <textarea id={label}
                    className={`${classes.input} ${classes.textarea}`}
                    value={value}
                    name={label}
                    placeholder={placeholder}
                    onChange={onChange}>
          </textarea>
          :
          <input id={label}
                 className={classes.input}
                 value={value}
                 name={label}
                 type={inputType}
                 placeholder={placeholder}
                 onChange={onChange}/>
      }
    </div>
  );
};

InputWithLabel.defaultProps = {
  inputType: 'text',
  className: '',
  placeholder: '',
  textarea: false
};

InputWithLabel.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  inputType: PropTypes.string,
  className: PropTypes.string,
  placeholder: PropTypes.string,
  textarea: PropTypes.bool,
  onChange: PropTypes.func.isRequired
};

export default InputWithLabel;
