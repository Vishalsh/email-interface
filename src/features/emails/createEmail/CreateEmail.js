import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'

import Button from 'components/button/Button';
import emailActions from 'features/emails/emails.actions';
import InputWithLabel from 'components/inputWithLabel/InputWithLabel';
import { isEmailValid } from 'utilities/formValidators';
import classes from './CreateEmail.module.scss';

export class CreateEmail extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: {
        to: '',
        cc: '',
        subject: '',
        body: ''
      }
    };
  }

  onChange = (event) => {
    const { name, value } = event.target;
    const { email } = this.state;

    this.setState({
      email: {
        ...email,
        [name]: value
      }
    })
  };

  isFormValid = () => {
    const { to } = this.state.email;

    return isEmailValid(to);
  };

  submit = (event) => {
    const { email } = this.state;
    event.preventDefault();

    this.props.sendEmail(email);
  };

  render() {
    const { toggleCreateEmailPopup } = this.props;
    const { email } = this.state;
    const { to, cc, subject, body } = email;

    return (
      <div className={classes.modalPopup}>
        <div className={classes.overlay}/>
        <section className={classes.createEmail}>
          <form>
            <InputWithLabel label="to" value={to} onChange={this.onChange}/>
            <InputWithLabel label="cc" value={cc} onChange={this.onChange}/>
            <InputWithLabel label="subject" value={subject} onChange={this.onChange}/>
            <InputWithLabel textarea label="body" value={body} onChange={this.onChange}/>

            <div className="row">
              <div className="col-sm-6">
                <Button type="primary"
                        className={classes.submitButton}
                        onClick={this.submit}
                        disabled={!this.isFormValid()}>
                  Send
                </Button>
              </div>
              <div className="col-sm-6">
                <Button type="default"
                        onClick={toggleCreateEmailPopup}>
                  Cancel
                </Button>
              </div>
            </div>
          </form>
        </section>
      </div>
    );
  }
}

CreateEmail.propTypes = {
  sendEmail: PropTypes.func.isRequired,
  toggleCreateEmailPopup: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => bindActionCreators({
  sendEmail: emailActions.sendEmail,
  toggleCreateEmailPopup: emailActions.toggleCreateEmailPopup
}, dispatch);

export default connect(null, mapDispatchToProps)(CreateEmail);
