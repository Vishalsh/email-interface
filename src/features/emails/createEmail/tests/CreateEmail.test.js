import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import InputWithLabel from 'components/inputWithLabel/InputWithLabel';
import Button from 'components/button/Button';
import { CreateEmail } from '../CreateEmail';

describe('CreateEmail', () => {
  const email = {
    to: 'msDhoni@bcci.com',
    cc: 'vKohli@bcci.com',
    subject: 'Lets Play',
    body: 'Lets Play the game today'
  };

  const props = {
    sendEmail: jest.fn(),
    toggleCreateEmailPopup: jest.fn()
  };
  const component = shallow(<CreateEmail {...props}/>);

  it('should render the component', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should set the email details', () => {
    component.find(InputWithLabel).get(0).props.onChange({ target: { name: 'to', value: email.to } });
    component.find(InputWithLabel).get(1).props.onChange({ target: { name: 'cc', value: email.cc } });
    component.find(InputWithLabel).get(1).props.onChange({ target: { name: 'subject', value: email.subject } });
    component.find(InputWithLabel).get(1).props.onChange({ target: { name: 'body', value: email.body } });

    expect(component.state().email.to).toEqual(email.to);
    expect(component.find(InputWithLabel).get(0).props.value).toEqual(email.to);
    expect(component.state().email.cc).toEqual(email.cc);
    expect(component.find(InputWithLabel).get(1).props.value).toEqual(email.cc);
    expect(component.state().email.subject).toEqual(email.subject);
    expect(component.find(InputWithLabel).get(2).props.value).toEqual(email.subject);
    expect(component.state().email.body).toEqual(email.body);
    expect(component.find(InputWithLabel).get(3).props.value).toEqual(email.body);
  });

  it('should disable the submit button when form is invalid', () => {
    component.setState({
      email: {
        to: 'foo@bar@com',
        cc: '',
        subject: 'Lets Play',
        body: ''
      }
    });

    expect(component.find(Button).get(0).props.disabled).toBeTruthy();
  });

  it('should send the email', () => {
    component.setState({
      email
    });
    const preventDefault = jest.fn();

    component.find(Button).get(0).props.onClick({ preventDefault });

    expect(preventDefault).toHaveBeenCalled();
    expect(props.sendEmail).toHaveBeenCalledWith(email);
  });

  it('should close the popup', () => {
    component.find(Button).get(1).props.onClick();

    expect(props.toggleCreateEmailPopup).toHaveBeenCalled();

    props.toggleCreateEmailPopup.mockReset();

    component.find('.overlay').props().onClick();

    expect(props.toggleCreateEmailPopup).toHaveBeenCalled();
  });
});