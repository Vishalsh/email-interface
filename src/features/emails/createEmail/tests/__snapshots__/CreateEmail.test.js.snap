// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEmail should render the component 1`] = `
<div
  className="modalPopup"
>
  <div
    className="overlay"
    onClick={[MockFunction]}
  />
  <section
    className="createEmail"
  >
    <form>
      <InputWithLabel
        className=""
        inputType="text"
        label="to"
        onChange={[Function]}
        placeholder=""
        textarea={false}
        value=""
      />
      <InputWithLabel
        className=""
        inputType="text"
        label="cc"
        onChange={[Function]}
        placeholder=""
        textarea={false}
        value=""
      />
      <InputWithLabel
        className=""
        inputType="text"
        label="subject"
        onChange={[Function]}
        placeholder=""
        textarea={false}
        value=""
      />
      <InputWithLabel
        className=""
        inputType="text"
        label="body"
        onChange={[Function]}
        placeholder=""
        textarea={true}
        value=""
      />
      <div
        className="row"
      >
        <div
          className="col-sm-6"
        >
          <Button
            className="submitButton"
            disabled={true}
            onClick={[Function]}
            type="primary"
          >
            Send
          </Button>
        </div>
        <div
          className="col-sm-6"
        >
          <Button
            className=""
            disabled={false}
            onClick={[MockFunction]}
            type="default"
          >
            Cancel
          </Button>
        </div>
      </div>
    </form>
  </section>
</div>
`;
